sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1160 @@
+        The non-singleton classes, encoded as integers in [0, n_classes).
+        If None (default), they will be inferred from ``y``.
+
+    **nn_kwargs : keyword arguments
+        Parameters to be passed to a :class:`neighbors.NearestNeighbors`
+        instance except from ``n_neighbors``.
+
+    Returns
+    -------
+    target_neighbors: array, shape (n_samples, n_neighbors)
+        The indices of the target neighbors of each training sample.
+    """
+
+    target_neighbors = np.zeros((X.shape[0], n_neighbors), dtype=np.intp)
+
+    nn = NearestNeighbors(n_neighbors=n_neighbors, **nn_kwargs)
+
+    if classes is None:
+        classes = np.unique(y)
+
+    for class_id in classes:
+        ind_class = np.where(y == class_id)[0]
+        nn.fit(X[ind_class])
+        neigh_ind = nn.kneighbors(return_distance=False)
+        target_neighbors[ind_class] = ind_class[neigh_ind]
+
+    return target_neighbors
+
+
+def _find_impostors_blockwise(X_a, X_b, radii_a, radii_b,
+                              return_distance=False, block_size=8):
+    """Find (sample, impostor) pairs in blocks to avoid large memory usage.
+
+    Parameters
+    ----------
+    X_a : array, shape (n_samples_a, n_components)
+        Transformed data samples from class A.
+
+    X_b : array, shape (n_samples_b, n_components)
+        Transformed data samples from class B.
+
+    radii_a : array, shape (n_samples_a,)
+        Squared distances of the samples in ``X_a`` to their margins.
+
+    radii_b : array, shape (n_samples_b,)
+        Squared distances of the samples in ``X_b`` to their margins.
+
+    block_size : int, optional (default=8)
+        The maximum number of mebibytes (MiB) of memory to use at a time for
+        calculating paired squared distances.
+
+    return_distance : bool, optional (default=False)
+        Whether to return the squared distances to the impostors.
+
+    Returns
+    -------
+    imp_indices : array, shape (n_impostors,)
+        Unraveled indices referring to a matrix of shape
+        (n_samples_a, n_samples_b). Index pair (i, j) is returned (
+        unraveled) if either sample i is an impostor to sample j or sample j
+        is an impostor to sample i.
+
+    imp_distances : array, shape (n_impostors,), optional
+        imp_distances[i] is the squared distance between samples imp_row[i] and
+        imp_col[i], where
+        imp_row, imp_col = np.unravel_index(imp_indices, (n_samples_a,
+        n_samples_b))
+    """
+
+    n_samples_a = X_a.shape[0]
+    bytes_per_row = X_b.shape[0] * X_b.itemsize
+    block_n_rows = int(block_size*1024*1024 // bytes_per_row)
+
+    imp_indices, imp_distances = [], []
+
+    # X_b squared norm stays constant, so pre-compute it to get a speed-up
+    X_b_norm_squared = row_norms(X_b, squared=True)[np.newaxis, :]
+    for chunk in gen_batches(n_samples_a, block_n_rows):
+        # The function `sklearn.metrics.pairwise.euclidean_distances` would
```

Comment:
scikit-learn's `pairwise_distances` has recently been a lot improved with chunk-wise computations and parallelism, maybe you can check how it compares with this computation (not only performance but also precision ?) Btw we probably need to use the parallelism option it in NCA I'll make a PR when I have time

Reply:
Will look into that.

Reply:
So, I tried to find if there is a way to use `pairwise_distances_chunked` but I think it's not possible in this case. Nevertheless, it's possible now to make use of `get_chunk_n_rows` which utilizes the user configuration for scikit learn for limited memory.